### PR TITLE
Ft/ MongoDB oplogs

### PR DESCRIPTION
**Pull request by JianqinWang**
JIRA ticket: https://scality.atlassian.net/browse/ZENKO-15
https://github.com/scality/Arsenal/pull/425 needs to be merged first.

### CODE REVIEW

File: lib/management.js [hunk 1]
```diff
@@ -3,12 +3,44 @@ const forge = require('node-forge');
...
-    host: config.queuePopulator.dmd.host,
-    port: config.queuePopulator.dmd.port,
-});
+const Logger = require('werelogs').Logger;
+
+const logger = new Logger('mongoclient');
+
+const serviceBucket = 'PENSIEVE';
+
+let setup = false;
```
Comment (bennettbuchanan) [inline lib/management.js]: I think it could be clearer to define this in the function `initManagement` where it is used and redefined.

File: extensions/replication/queueProcessor/task.js [hunk 2]
```diff
@@ -35,7 +35,6 @@ function initAndStart() {
             setTimeout(initAndStart, 5000);
             return;
         }
-        console.log('management init done');
         queueProcessor.start();
         // setInterval(initManagement, 5000, error => {
```
Comment (bennettbuchanan) [inline extensions/replication/queueProcessor/task.js]: Since moving the `setInterval` to `initManagement` maybe we can go ahead and remove this comment?

File: lib/management.js [hunk 3]
```diff
@@ -245,64 +290,72 @@ function saveServiceCredentials(conf, auth) {
...
+        return async.waterfall([
+            cb => {
+                if (!setup) {
+                    setup = true;
+                    metadata = new Metadata('mongodb', params, logger);
+                    return metadata.setup(() => cb());
+                }
+                return process.nextTick(cb());
+            },
+            cb => {
```
Comment (bennettbuchanan) [inline lib/management.js]: Minor: for these steps in the waterfall we could save some lines by not using the braces (as done below https://github.com/scality/backbeat/pull/187/files#diff-c529f19cba932673dfc25817365dec76R334).

File: lib/management.js [hunk 4]
```diff
@@ -3,12 +3,44 @@ const forge = require('node-forge');
...
-
-const MetadataFileClient = arsenal.storage.metadata.MetadataFileClient;
-const mdClient = new MetadataFileClient({
-    host: config.queuePopulator.dmd.host,
-    port: config.queuePopulator.dmd.port,
-});
+const Logger = require('werelogs').Logger;
+
+const logger = new Logger('mongoclient');
+
```
Comment (bennettbuchanan) [inline lib/management.js]: It could be useful to add a configuation to werelogs so we can manage the log level from the config.json, e.g.:

```js
werelogs.configure({ 
    level: config.log.logLevel,
    dump: config.log.dumpLevel,
});
```

File: conf/config.json [hunk 5]
```diff
@@ -70,7 +78,7 @@
         }
     },
     "log": {
-        "logLevel": "info",
+        "logLevel": "trace",
```
Comment (bennettbuchanan) [inline conf/config.json]: Do we want to set the log level to trace?

File: lib/credentials/AccountCredentials.js [hunk 6]
```diff
@@ -66,14 +72,15 @@ class StaticFileAccountCredentials extends AccountCredentials {
...
         assert.strictEqual(authConfig.type, 'service');
-        super(authConfig, log);
+        const serviceAuth = management.getLatestServiceAccountCredentials();
+        super(authConfig, log, serviceAuth);
     }
 
-    initAuthData() {
-        this._authdata = management.getLatestServiceAccountCredentials();
-    }
+    // initAuthData() {
```
Comment (bennettbuchanan) [inline lib/credentials/AccountCredentials.js]: Is this okay to be removed?

File: lib/management.js [hunk 7]
```diff
@@ -148,59 +183,65 @@ function applyReplicationState(conf, currentStateSerialized, db, cb) {
...
                 putReplication(stream, err => {
                     if (err) {
                         return cb(err);
                     }
                     return streamConfigured(stream, cb);
                 });
             } else {
-                console.log('stream has expected state, doing nothing', stream.streamId, stream.enabled);
+                logger.debug('stream has expected state, doing nothing' +
+                    `${stream.streamId}, ${stream.enabled}`);
```
[the remark this answers is not in the document]
Comment (JianqinWang) [reply lib/management.js]: I asked around and I think people told me that it's ok to not include an object/second arg for the logger - I'm also just converting the previous console.logs, so I don't want to add too much unnecessary info

[the remark this answers is not in the document]
Comment (bennettbuchanan) [reply lib/management.js]: Oh yes, this looks fine. I had seen the comma without realizing it was part of the template string.

File: Dockerfile [hunk 8]
```diff
@@ -2,12 +2,13 @@ FROM node:6-slim
...
+RUN apt-get update \
+    && apt-get install -y jq python git build-essential --no-install-recommends
+
 # Keep the .git directory in order to properly report version
 COPY . /usr/src/app
 
-RUN apt-get update \
-    && apt-get install -y jq python git build-essential --no-install-recommends \
-    && npm install --production \
+RUN npm install --production \
```
Comment (rahulreddy) [inline Dockerfile]: What is the reason for this change? Doesn't this add an extra layer to the image?

File: lib/management.js [hunk 9]
```diff
@@ -3,12 +3,46 @@ const forge = require('node-forge');
...
 
-const MetadataFileClient = arsenal.storage.metadata.MetadataFileClient;
-const mdClient = new MetadataFileClient({
-    host: config.queuePopulator.dmd.host,
-    port: config.queuePopulator.dmd.port,
+werelogs.configure({
+    level: config.log.logLevel,
+    dump: config.log.dumpLevel,
 });
+const logger = new werelogs.Logger('mongoclient');
```
Comment (rahulreddy) [inline lib/management.js]: Since this is for all backends (Metadata, Mongo), we should make it more general `MetadataManager` or `Management`, I am ok with either.

File: lib/management.js [hunk 10]
```diff
@@ -3,12 +3,46 @@ const forge = require('node-forge');
...
+
+const serviceBucket = 'PENSIEVE';
+
+const Metadata = arsenal.storage.metadata.MetadataWrapper;
+let metadata;
+
+const params = {
```
Comment (rahulreddy) [inline lib/management.js]: I am assuming that we are planning on cleaning this up later.

File: lib/management.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (rahulreddy) [reply lib/management.js]: We wrote this down in logging guidelines so that everyone writes logs the same way https://github.com/scality/Guidelines/blob/master/CONTRIBUTING.md#examples

File: lib/management.js [hunk 11]
```diff
@@ -148,59 +188,66 @@ function applyReplicationState(conf, currentStateSerialized, db, cb) {
...
+                logger.debug(`enabling new stream ${stream.streamId}`);
                 putReplication(stream, err => {
                     if (err) {
                         return cb(err);
                     }
                     return streamConfigured(stream, cb);
                 });
             } else {
-                console.log('new disabled stream, doing nothing', stream.streamId);
+                logger.log('new disabled stream, doing nothing' +
```
Comment (rahulreddy) [inline lib/management.js]: `logger.debug` you mean?

File: lib/credentials/AccountCredentials.js [hunk 12]
```diff
@@ -2,14 +2,20 @@ const assert = require('assert');
...
+
+const authdata = require('../../conf/authdata.json');
 
 class AccountCredentials extends AWS.Credentials {
-    constructor(authConfig, log) {
-        this._log = log;
-        this.initAuthData();
-        const accountInfo = this.getAuthData().accounts.find(
-            account => account.name === authConfig.account);
+    constructor(authConfig, log, serviceAuth) {
```
Comment (RachedBenMustapha) [inline lib/credentials/AccountCredentials.js]: Why move the provisioned service account logic here? it makes `ProvisionedServiceAccountCredentials` basically a passthrough while the base class now has knowledge of several auth types

File: lib/management.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (RachedBenMustapha) [reply lib/management.js]: Agreed, since you are just converting it's fine, but for werelogs logs the message should not have "content", content should go in an attached object to be searchable.

File: lib/management.js (hunk 11 above)
Comment (RachedBenMustapha) [inline lib/management.js]: This probably crashes

File: lib/queuePopulator/LogReader.js [hunk 13]
```diff
@@ -252,6 +252,7 @@ class LogReader {
             readOptions.limit = params.maxRead;
         }
         this.log.debug('reading records', { readOptions });
+        // need to implement this method for mongo!!
```
Comment (RachedBenMustapha) [inline lib/queuePopulator/LogReader.js]: what happens if it's not implemented?

File: lib/management.js [hunk 14]
```diff
@@ -245,64 +297,72 @@ function saveServiceCredentials(conf, auth) {
...
+            cb => {
+                if (!setup) {
+                    setup = true;
+                    metadata = new Metadata('mongodb', params, bucketclient,
+                        logger);
+                    return metadata.setup(() => cb());
+                }
+                return process.nextTick(cb);
+            },
+            cb => metadata.listObject(serviceBucket, {}, logger, () => cb()),
```
Comment (RachedBenMustapha) [inline lib/management.js]: Why list objects here?

File: lib/management.js [hunk 15]
```diff
@@ -245,64 +297,72 @@ function saveServiceCredentials(conf, auth) {
...
+            cb => {
+                metadata.getObjectMD(serviceBucket,
+                'configuration/overlay-version', {}, logger, (err, res) =>
+                    cb(err, res));
+            },
+            (version, cb) => {
```
Comment (RachedBenMustapha) [inline lib/management.js]: could be written as `(version, cb) => loadOverlayVersion(metadata, version, cb)`, as this waterfall is becoming busy

File: Dockerfile (hunk 8 above)
Comment (JianqinWang) [reply Dockerfile]: I think @ssalaues mentioned that this would speed up the docker build process?

Comment (ssalaues) [reply Dockerfile]: Yeah it's a faster docker build because the apt update and apt install doesn't ever change so the the builds can use that cached layer. It's not at all critical to the PR but helped speed up the feedback loop during build and testing.

Comment (RachedBenMustapha) [reply Dockerfile]: Yeah extra layers actually are a good thing when placed strategically, not sure why they have a bad rep ;)

File: lib/queuePopulator/LogReader.js (hunk 13 above)
Comment (JianqinWang) [reply lib/queuePopulator/LogReader.js]: Mmmmm I think this was a residual comment before the completion of this: https://github.com/scality/Arsenal/blob/master/lib/storage/metadata/mongoclient/LogConsumer.js
:sweat_smile:

**Comment by JianqinWang** [comment]
@ironman-machine try
